CMakeLists.txt

```diff
@@ -544,6 +544,7 @@ target_link_libraries(torrent-rasterbar
 if (WIN32)
 	target_link_libraries(torrent-rasterbar
 		PRIVATE
+			$<$<NOT:$<BOOL:${BUILD_SHARED_LIBS}>>:crypt32>
```

Comment:
PRIVATE? How do this propagates to the test executables? Does this change fix the problem?

Reply:
It does work, yes. I see in the output dump of the compiler/linker commands that when `BUILD_SHARED_LIBS` is true, `crypt32` is not linked, and when it is false, `crypt32` gets linked.

Reply:
Linked to the examples?

Reply:
Yes, both examples and tools.

Reply:
Must be something special about static libs on Windows. Can't understand both the problem and the solution :D

Reply:
@zeule my initial solution was an attempt to solve this problem: https://github.com/microsoft/vcpkg/pull/10686#issuecomment-612817769. Then I simply made the change you suggested which makes sense, as it's doing the same thing but "upstream". However, I also don't understand why _only_ static builds would need this either. Could it be the case that this is only needed if using static libcrypto? I ask this because this happened with `vcpkg`, and a static build there pulls in static-everything.

Reply:
…> static build there pulls in static-everything.

Aha, thanks! that helps to understand it a bit better. Seems like there is no need to make changes to libtorrent target. You might need to play with `OPENSSL_USE_STATIC_LIBS` and `OPENSSL_MSVC_STATIC_RT` in vcpkg, or it's a bug in cmake (FindOpenSSL module). Please check FindOpenSSL.cmake for details.

Reply:
Wait, we have them already under the `static_runtime` flag. Doe vcpkg sets this option?

Reply:
> Does vcpkg set this option?

Not sure, but I don't think so. I will check that.

Reply:
There is no "crypt32" string in the FindOpenSSL.cmake on my machine. The log in microsoft/vcpkg#10686 shows it is the OpenSSL libcrypto who needs crypt32.lib. Here is OpenSSL issue: https://github.com/openssl/openssl/issues/1061. I'd you encountered a  bug in FindOpenSSL.cmake.

Reply:
Finally the root of it: https://gitlab.kitware.com/cmake/cmake/-/issues/19263

Reply:
@zeule nice find. I haven't yet checkedif passing in `static_runtime` fixes the issue, but if I understand correctly, it won't, right?

Reply:
Yes, it won't. You need to patch FindOpenSSL.cmake instead.

Reply:
@zeule Until the problem in in `FindOpenSSL.cmake` is fixed upstream, how about I add a comment like `TODO: needed until https://gitlab.kitware.com/cmake/cmake/-/issues/19263 is fixed` next to the change?

Reply:
Me? It's up to @arvidn.

Reply:
If put the patch here, here is my suggestion: 
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index c8e284480..5f1fa52c5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -262,6 +262,7 @@ if (encryption)
    TYPE RECOMMENDED
    PURPOSE "Provides encryption support to libtorrent"
  )
+ target_link_libraries(OpenSSL::SSL INTERFACE $<$<BOOL:${OPENSSL_USE_STATIC_LIBS}>:crypt32>)
  target_sources(torrent-rasterbar PRIVATE src/mpi src/pe_crypto)
  target_link_libraries(torrent-rasterbar PUBLIC OpenSSL::SSL)
  target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_OPENSSL)
```

Reply:
I forgot to add the `WIN32` condition, but the idea is clear, I believe

Reply:
> If put the patch here, here is my suggestion:
> 
> ...

What is the advantage of doing it that way?

Also personally I'm not a fan of putting the `WIN32` condition in the generator expression as well, it will become really unreadable, but if it is technically the much better option, well...

Reply:
We went from unconditionally appending crypt32 to the executables to fixing a bug in cmake modifying the target which is has incorrect dependencies.

Reply:
@zeule I noticed in the CMake file, `OPENSSL_USE_STATIC_LIBS` only gets set to `TRUE` if `static-runtime` is used, but not if `BUILD_SHARED_LIBS` is false; I don't think that's what we want, because then if `BUILD_SHARED_LIBS` is false but `static-runtime` is not specified, the build will still fail.

Reply:
Pardon me, but it can't be true. Whether a build fails depends only on the OpenSSL libraries. FindOpenSSL.cmake does not account for BUILD_SHARED_LIBS. If there are only static OpenSSL libs available, the libtorrent build will fail always.

Reply:
@zeule which `CMakeLists.txt` file is that patch for? I grep'd "Provides encryption support to libtorrent" but found nothing.

Reply:
Sorry, that was RC_1_1 branch. [Here](https://github.com/arvidn/libtorrent/blob/RC_1_2/CMakeLists.txt#L636)

Reply:
@zeule 
Still, I don't understand why you suggest `if (WIN32 AND OPENSSL_USE_STATIC_LIBS)` and not `if (WIN32 AND BUILD_SHARED_LIBS)`. The build fails with the former, unless I pass in `static_runtime`, whereas with the latter, the build succeeds regardless of `static_runtime`.

Reply:
This is puzzling. Could you, please, share contents of the `INTERFACE_LINK_LIBRARIES` property for the `OpenSSL::SSL` target in both cases?

```cmake
get_property(ill TARGET OpenSSL::SSL PROPERTY INTERFACE_LINK_LIBRARIES)
message(STATUS "ill: ${ill}")
```

Reply:
And `IMPORTED_LOCATION` too, please.

Reply:
@zeule 
I appended this at the end of the `CMakeLists.txt`:

```
get_property(ill TARGET OpenSSL::SSL PROPERTY INTERFACE_LINK_LIBRARIES)
message(STATUS "ill: ${ill}")
get_property(imp TARGET OpenSSL::SSL PROPERTY IMPORTED_LOCATION)
message(STATUS "imp: ${imp}")
```

This was the output at the end of the configuring step (in `vcpkg` this goes to a log file), building `libtorrent[examples,tools]:x64-windows-static` from my `vcpkg` PR, except with your patch from https://github.com/arvidn/libtorrent/pull/4547#discussion_r409779283 instead of my "force link"`crypt32` patches:
```
ill: OpenSLL::Crypto
itt: C:/Users/User/Documents/vcpkg/installed/x64-windows-static/lib/libssl.lib
```

Reply:
@zeule 
I guess I can just pass `-DOPENSSL_USE_STATIC_LIBS=ON` manually to the libtorrent build. That way I can have that without using `-Dstatic_runtime`. Was this what you meant?

Reply:
No, that is not what I meant. You told us the libtorrent build fails if you define `BUILD_SHARED_LIBS` to `False`. But the linking error comes from `libcrypt` (OpenSLL::Crypto). How that one can depend on `BUILD_SHARED_LIBS`? If we want to understand what's going on, we need `INTERFACE_LINK_LIBRARIES` and `IMPORTED_LOCATION` for OpenSLL::Crypto in the vcpkg environment for these three cases, please:

1. `BUILD_SHARED_LIBS=True`, `OPENSSL_USE_STATIC_LIBS=False`
2. `BUILD_SHARED_LIBS=False`, `OPENSSL_USE_STATIC_LIBS=False`
3. `BUILD_SHARED_LIBS=False`, `OPENSSL_USE_STATIC_LIBS=True`

Reply:
Actually, thinking about it a bit more, I think the correct thing to do is to just pass `-DOPENSSL_USE_STATIC_LIBS=ON` if the library linkage is static, but not the CRT linkage, and if both are static, just pass `-Dstatic_runtime`.

Reply:
> Actually, thinking about it a bit more, I think the correct thing to do is to just pass `-DOPENSSL_USE_STATIC_LIBS=ON` if the library linkage is static, but not the CRT linkage, and if both are static, just pass `-Dstatic_runtime`.

Please note I wrote this comment before seeing your previous (the conversation did not update on my end for some reason).

Reply:
> You told us the libtorrent build fails if you define `BUILD_SHARED_LIBS` to `False`

No, I said that's when the build succeeded (because when building `x64-windows-static`, `BUILD_SHARED_LIBS` gets set to `FALSE`) with the previous patch (`$<$<NOT:$<BOOL:${BUILD_SHARED_LIBS}>>:crypt32>`). In the same circusmstances, the build was failing with your patch:

```
if (WIN32 AND OPENSSL_USE_STATIC_LIBS)
    target_link_libraries(OpenSSL::SSL INTERFACE crypt32)
endif()
```

Reply:
> think the correct thing to do is to just pass -DOPENSSL_USE_STATIC_LIBS=ON

But that should not help, because of the bug in CMake-supplied OpenSSL module, right?

Reply:
@zeule 
The whole point of setting `OPENSSL_USE_STATIC_LIBS` is to satisfy the condition imposed by your patch (https://github.com/arvidn/libtorrent/pull/4547#discussion_r409779283), which is now what the PR is.

The way the PR is now, everything seems to work correctly.

- static library linkage + static CRT -> set `-Dstatic_runtime=ON`, build succeeds
- dynamic library linkage + dynamic CRT -> do nothing, build succeeds
- static library linkage + dynamic CRT -> set `-DOPENSSL_USE_STATIC_LIBS=ON`, build succeeds.

Reply:
I suppose this is the correct thing to do, the OpenSSL docs have this to say:

> If you **link with static OpenSSL libraries then you're expected to
 additionally link** your application with WS2_32.LIB, GDI32.LIB,
 ADVAPI32.LIB, **CRYPT32.LIB** and USER32.LIB.

(emphasis mine, source: https://github.com/openssl/openssl/blob/6f892296038490a7fa24b32ac6f7305687634fb0/NOTES.WIN#L127)

which is what this patch achieves now, provided the user passes `-Dstatic_runtime=ON` or `-DOPENSSL_USE_STATIC_LIBS=ON`, depending on what they need.

Reply:
Actually, I just noticed that currently `advapi32` is not mentioned anywhere, and it is another library that it is recommended to link against (ignoring the GUI-related libraries). Currently, there are no build errors due to the lack of it, but it could be needed due to some future change. @arvidn @zeule is it needed?